### DISCUSSION TITLE
Usage Graphene Subscriptions in FastAPI - Usage of Starlette-Graphene3 failed

### DISCUSSION THREAD

**Discussion by Sharleedah** [fastapi/fastapi#6922] [Questions]
I wanted to make the subscriptions of graphene work in FastAPI. Unfortuantely I saw two other repositories who asked a similar thing and no conclusion made there. After some research I found the package starlette-graphene3, which makes uploading files and the subscription possible in Starlette. Due to the usage of the the GraphQLApp-class of starlette i thought, that this is going to work. Unfortunately there is a little problem when using starlette-graphene3 with fastapi:

- When installing starlette-graphene3 I get the error: `ERROR: fastapi 0.63.0 has requirement starlette==0.13.6, but you'll have starlette 0.14.1 which is incompatible.`, which makes FastAPI not work anymore.

My questions are: 
- is it in planning to use a newer Starlette version, that starlette-graphene3 could be used
- Any other way to make subscriptions work in FastAPI?

When trying to use GraphQL  like in the documentation of FastAPI with subscriptions i get the error:
`"Subscriptions are not allowed. You will need to either use the subscribe function or pass allow_subscriptions=True"`

Code to reproduce:
```Python
from fastapi import FastAPI
from starlette.graphql import GraphQLApp
import graphene
import asyncio

class Subscription(graphene.ObjectType):
    count = graphene.Int(upto=graphene.Int())

    async def subscribe_count(root, info, upto=3):
        for i in range(upto):
            yield i
            await asyncio.sleep(1)


app = FastAPI()
app.add_route("/graphql", GraphQLApp(schema=graphene.Schema(subscription=Subscription)))
```

**Comment by Mause** [comment]
For the reasoning behind pinning starlette to a specific version, see my comment here: https://github.com/tiangolo/fastapi/discussions/2532#discussioncomment-229069

As far as your question goes, what happens if you try the suggestions in the error regarding allowing subscriptions?

**Comment by Sharleedah** [comment]
Thanks for the fast response! Unfortunately I couldnt find the location where i can set this. The only solutions i found where directly dedicated to django. Do you have an idea where to set this setting?

**Comment by Sharleedah** [comment]
We can close this issue. I got it to work with Ariadne

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! :clap: :bow: 

Thanks for reporting back and closing the issue @Sharleedah  :+1:

**Comment by ppbpdx** [comment]
> We can close this issue. I got it to work with Ariadne

Any chance you can share the approach? I have been starting to research subscriptions for graphQL and python, specifically for FastApi and so far it looks as if the answers are pretty sparse. I'd love to help document how do implement them when we figure out a good pattern.

**Comment by ppbpdx** [comment]
> We can close this issue. I got it to work with Ariadne
@Sharleedah 

Would you be willing to share you approach with Ariadne? I seem to have it 80% of the way there but it is unclear to me how to expose the fastapi socket such that the subscription can be accessed through the socket. It seems as if I need to have a socket endpoint but it is not jumping out at me from the examples how to do this within fastapi.

Right now I get an error that complains about :
`{
  "error": "Could not connect to websocket endpoint ws://0.0.0.0:8000/graphql/. Please check if the endpoint url is correct."
}`

My app looks like this:

```

import asyncio
import uvicorn
from fastapi import FastAPI
from ariadne import  make_executable_schema, SubscriptionType
from ariadne.asgi import GraphQL

app = FastAPI()

# ariadne copied from https://ariadnegraphql.org/docs/subscriptions

type_def = """
    type Query {
        _unused: Boolean
    }

    type Subscription {
        counter: Int!
    }
"""

subscription = SubscriptionType()

@subscription.source("counter")
async def counter_generator(obj, info):
    for i in range(5):
        await asyncio.sleep(1)
        yield i

@subscription.field("counter")
def counter_resolver(count, info):
    return count + 1


schema = make_executable_schema(type_def, subscription)

app.mount("/graphql", GraphQL(schema, debug=True))


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

@tiangolo If I can work out a good example I will volunteer to write up a graphql subcription addition to the fastapi docs for your consideration.

**Comment by TheDelus** [comment]
> > We can close this issue. I got it to work with Ariadne
> > @Sharleedah
> 
> Would you be willing to share you approach with Ariadne? I seem to have it 80% of the way there but it is unclear to me how to expose the fastapi socket such that the subscription can be accessed through the socket. It seems as if I need to have a socket endpoint but it is not jumping out at me from the examples how to do this within fastapi.
> 
> Right now I get an error that complains about :
> `{ "error": "Could not connect to websocket endpoint ws://0.0.0.0:8000/graphql/. Please check if the endpoint url is correct." }`
> 
> My app looks like this:
> 
> ```
> 
> import asyncio
> import uvicorn
> from fastapi import FastAPI
> from ariadne import  make_executable_schema, SubscriptionType
> from ariadne.asgi import GraphQL
> 
> app = FastAPI()
> 
> # ariadne copied from https://ariadnegraphql.org/docs/subscriptions
> 
> type_def = """
>     type Query {
>         _unused: Boolean
>     }
> 
>     type Subscription {
>         counter: Int!
>     }
> """
> 
> subscription = SubscriptionType()
> 
> @subscription.source("counter")
> async def counter_generator(obj, info):
>     for i in range(5):
>         await asyncio.sleep(1)
>         yield i
> 
> @subscription.field("counter")
> def counter_resolver(count, info):
>     return count + 1
> 
> 
> schema = make_executable_schema(type_def, subscription)
> 
> app.mount("/graphql", GraphQL(schema, debug=True))
> 
> 
> if __name__ == "__main__":
>     uvicorn.run(app, host="0.0.0.0", port=8000)
> ```
> 
> @tiangolo If I can work out a good example I will volunteer to write up a graphql subcription addition to the fastapi docs for your consideration.

Your code works for me. Have you installed uvicorn with extras? See https://github.com/tiangolo/fastapi/issues/2209#issuecomment-713341164

**Comment by matthewchung74** [comment]
@TheDelus any luck? I would love to see a graphql subscription addition!

**Comment by TheDelus** [comment]
@matthewchung74 yes with Ardiadne + FastAPI + uvicorn with extras I got it working. I tested it with GraphiQL and [urql](https://github.com/FormidableLabs/urql). The code from @ppbpdx is working.

**Comment by matthewchung74** [comment]
@TheDelus awesome. Just so I'm clear, it's the code where @ppbpdx says "My app looks like this:" right?

**Comment by TheDelus** [comment]
@matthewchung74 Yep thats it

**Comment by aliscie** [comment]
when `pip install ariadne` I got 
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
graphql-relay 2.0.1 requires graphql-core<3,>=2.2, but you have graphql-core 3.1.6 which is incompatible.
graphene 2.1.8 requires graphql-core<3,>=2.1, but you have graphql-core 3.1.6 which is incompatible.
```

**Comment by yezz123** [comment]
> when `pip install ariadne` I got
> 
> ```
> ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
> graphql-relay 2.0.1 requires graphql-core<3,>=2.2, but you have graphql-core 3.1.6 which is incompatible.
> graphene 2.1.8 requires graphql-core<3,>=2.1, but you have graphql-core 3.1.6 which is incompatible.
> ```

Uninstall the graphene and graphql-core dependency, then re-install `pip install ariadne`, that relate to a conflict between the versions.

**Comment by aliscie** [comment]
I `pip uninstall graphql-relay graphene graphql-core ariadne` then `pip install ariadne` but I got error `ModuleNotFoundError: No module named 'graphene'` and when I install `graphene` I got `ariadne 0.13.0 requires graphql-core>=3.1.0, but you have graphql-core 2.3.2 which is incompatible.` again 

# that is a very weird closed loop

**Comment by yezz123** [comment]
> I `pip uninstall graphql-relay graphene graphql-core ariadne` then `pip install ariadne` but I got error `ModuleNotFoundError: No module named 'graphene'` and when I install `graphene` I got `ariadne 0.13.0 requires graphql-core>=3.1.0, but you have graphql-core 2.3.2 which is incompatible.` again
> 
> # that is a very weird closed loop

Try This :
- `pip install graphql-core==3.0.5`
- `pip install graphene`
- `pip install ariadne`

**Comment by aliscie** [comment]
you mean `pip install graphene`!
`graphene 2.1.9 requires graphql-core<3,>=2.1, but you have graphql-core 3.1.6 which is incompatible.`

**Comment by yezz123** [comment]
> you mean `pip install graphene`!
> `graphene 2.1.9 requires graphql-core<3,>=2.1, but you have graphql-core 3.1.6 which is incompatible.`

Literally that weird, cause when we prepare the environment we put all the requirements.txt like that :
```
fastapi
graphene
ariadne
```
you can give it a try in this way `pip install -r requirements.txt`

**Comment by aliscie** [comment]
ok now it worked, but still weird 😂.
However, can i do `app.add_route("/graphql", GraphQLApp(schema=graphene.Schema(query=Query, mutation=PostMutations, subscription=GraphQL(schema, debug=True))))` in order to have all methods?
in gernall **How to have Query, mutation and subscription methods at ones**?

**Comment by yezz123** [comment]
@aliscie you can check the documentation to get the main basics or there is some useful links that help me to create [JeffQL](https://github.com/yezz123/JeffQL).

- https://fastapi.tiangolo.com/advanced/graphql/
- http://blog.adnansiddiqi.me/getting-started-with-graphql-in-python-with-fastapi-and-graphene/
- https://blog.logrocket.com/building-a-graphql-server-with-fastapi/
- https://testdriven.io/blog/fastapi-graphql/
- https://dev.to/bekbrace/graphql-and-fastapi-combination-1c37
- https://www.youtube.com/watch?v=Puvr82Cm26o&ab_channel=VeryAcademy

**Comment by aliscie** [comment]
I know all of them it is just painful to not to find a scaffold with all the methods at ones 😔.

**Comment by aliscie** [comment]
solved this is a template for queries, mutatoins and subscriptions with testing everything in one https://github.com/aliscie/fastapi-full-graphql-template
